File: test/bench/cpp/nqueens.cpp
// NQueens solution in C++
// Note: does not free memory as that is difficult to do
// since many subsolutions are shared
#include <iostream>

template <typename T>
class list {
public:
  T head;
  list<T>* tail;
  list(T hd, list<T>* tl) {
    head = hd;
    tail = tl;
  }
  ~list() {
    delete head;
    delete tail;
  }
};

template <typename T>
list<T>* Cons( T hd, list<T>* tl ) {
  return new list<T>(hd,tl);
}

template <typename T>
int len(list<T>* xs) {
  int n = 0;
  while(xs != NULL) {
    n++;
    xs = xs->tail;
  }
  return n;
}

bool safe( int queen, list<int>* xs ) {
  list<int>* cur = xs;
  int diag = 1;
  while(cur != NULL) {
    int q = cur->head;
    if (queen == q || queen == (q+diag) || queen == (q-diag)) {
      return false;
    }
    diag++;
    cur = cur->tail;
  }
  return true;
}

list<list<int>*>* append_safe( int k, list<int>* soln, list<list<int>*>* solns ) {
  list<list<int>*>* acc = solns;
  int n = k;
  while(n > 0) {
    if (safe(n,soln)) {
      acc = Cons(Cons(n,soln),acc);
    }
    n--;
  }
  return acc;
}

list<list<int>*>* extend( int n, list<list<int>*>* solns ) {
  list<list<int>*>* acc = NULL;
  list<list<int>*>* cur = solns;
  while(cur != NULL) {
    list<int>* soln = cur->head;
    acc = append_safe(n,soln,acc);
    cur = cur->tail;
  }
  return acc;
}

list<list<int>*>* find_solutions( int n ) {
  int k = 0;
  list<list<int>*>* acc = Cons<list<int>*>(NULL,NULL);
  while(k < n) {
    acc = extend(n,acc);
    k++;
  }
  return acc;
}

int nqueens(int n) {
  return len(find_solutions(n));
}

int main(int argc, char ** argv) {
    int n = 13;
    if (argc == 2) {
      n = atoi(argv[1]);
    }
    std::cout << nqueens(n) << "\n";
    return 0;
}

File: test/bench/java/nqueens.java
class List<T> {
  T head;
  List<T> tail;

  List(T h, List<T> t) {
    head = h;
    tail = t;
  }

  static <T> int len( List<T> xs ) {
    int n = 0;
    while(xs != null) {
      n++;
      xs = xs.tail;
    }
    return n;
  }

  static<T> List<T> Cons( T h, List<T> t ) {
    return new List<T>(h,t);
  }

}

public class nqueens {
  static boolean safe( int queen, List<Integer> xs ) {
    int diag = 1;
    while(xs != null) {
      int q = xs.head;
      if (queen == q || queen == (q + diag) || queen == (q - diag)) {
        return false;
      }
      diag++;
      xs = xs.tail;
    }
    return true;
  }

  static List<List<Integer>> appendSafe( int k, List<Integer> soln, List<List<Integer>> solns ) {
    List<List<Integer>> acc = solns;
    while(k > 0) {
      if (safe(k,soln)) {
        acc = List.Cons(List.Cons(k,soln),acc);
      }
      k--;
    }
    return acc;
  }

  static List<List<Integer>> extend( int n, List<List<Integer>> solns ) {
    List<List<Integer>> acc = null;
    List<List<Integer>> cur = solns;
    while(cur != null) {
      acc = appendSafe(n, cur.head, acc);
      cur = cur.tail;
    }
    return acc;
  }

  static List<List<Integer>> findSolutions( int n ) {
    int k = 0;
    List<List<Integer>> acc = List.Cons(null,null);
    while(k < n) {
      acc = extend(n,acc);
      k++;
    }
    return acc;
  }


  static int nqueens(int n) {
    return List.len(findSolutions(n));
  }
  public static void main(String args[])
  {
      System.out.println( nqueens(13) );
  }
}
